Advanced search

wpf500 opened the pull request:
Implementation of advanced search!

![image](https://user-images.githubusercontent.com/2084823/201700091-e02924e1-6072-438e-a454-0d96d39433cb.png)

Changes from the design spec:
* Hiding save segment for now (is in the next piece of work)
* Added paginated result display (count and pages) above table
* Removed Cancel button in favour of Reset which resets the query to either the current segment or all results
* Moved Search button to the left as without Save Segment it looked strange by itself and we have a UI pattern of "<Action> | <Cancel>" (in this case Search | Reset)

As part of this work there have been a few improvements to components:
* `AppSelect` now supports option groups
* `AppInput` now supports a prefix (e.g. €)
* `AppRadioGroup` has better alignment, supports validation and can auto generate `name` attribute values
* Introduced `AppPaginatedResult` for a consistent UI for `Paginated` tables

wpf500 committed:
chore: add validation and input classes to select input

wpf500 committed:
feat: start of advanced search

wpf500 committed:
chore: unify styles between form components

wpf500 committed:
feat: lots of work on advanced search, add i18n

wpf500 committed:
feat: convert paginated rules into filters and vice versa

wpf500 committed:
feat: add filter summary

wpf500 committed:
feat: render operator args correctly in readonly display

wpf500 committed:
chore: add option to hide error message on AppInput

wpf500 committed:
Merge branch 'main' into feat/advanced-search

wpf500 committed:
Merge branch 'main' into feat/advanced-search

wpf500 committed:
Merge branch 'main' into feat/advanced-search

wpf500 committed:
fix: eslint problems and shared ref to filters list

wpf500 committed:
fix: all contacts count updating on search

wpf500 committed:
feat: initial support for vastly more filters and operators

wpf500 committed:
feat: add inclusive toggle

wpf500 committed:
feat: move SearchBox to more general component AppSearch

wpf500 committed:
feat: add support for required check to AppRadioGroup

wpf500 committed:
fix: updating filter

wpf500 committed:
chore: make params required

wpf500 committed:
chore: add unique name generator for AppRadioGroup

wpf500 committed:
feat: define labels in filter definition

wpf500 committed:
feat: add input for relative/absolute date toggle

wpf500 committed:
fix: define values for args

wpf500 committed:
fix: size changing between relative and text inputs

wpf500 committed:
feat: remove inclusive switch in favour of all operators in drop down

wpf500 committed:
chore: get currency symbol from API

wpf500 committed:
feat: support prefix in AppInput

wpf500 committed:
feat: lots of i18n

wpf500 committed:
fix: alignment of inline radio group

wpf500 committed:
fix: make duplicate of values to avoid mutation

wpf500 committed:
refactor: clean up where code should be

wpf500 committed:
chore: update i18n

wpf500 committed:
fix: error state in app input

wpf500 committed:
Merge branch 'main' into feat/advanced-search

wpf500 committed:
chore: show results count

wpf500 committed:
feat: add remove filter button

wpf500 committed:
fix: heading for search

wpf500 committed:
refactor: remove EmptyFilter in favour of simpler logic

wpf500 committed:
fix: layout bugs

wpf500 committed:
chore: make tags nullable

wpf500 committed:
chore: use beabee-common

wpf500 committed:
chore: update locales

wpf500 committed:
feat: add support for optgroups

wpf500 committed:
refactor: first pass at using beabee-common

wpf500 committed:
refactor: use beabee-common with type safety

wpf500 committed:
feat: bump beabee-common

wpf500 committed:
chore: removed unnecessary ?

wpf500 committed:
chore: migrate a few more data types to beabee-common

wpf500 committed:
Merge branch 'main' into feat/advanced-search

wpf500 committed:
chore: update i18n

wpf500 committed:
refactor: upgrade to beabee-common 1.7.3

wpf500 committed:
chore: add contributionCancelled

wpf500 committed:
chore: don't filter for empty search terms

wpf500 committed:
fix: properly set initial filter based on type

wpf500 committed:
Merge branch 'main' into feat/advanced-search

wpf500 committed:
fix: build errors

wpf500 committed:
fix: create new defaults every time to avoid reference problems

wpf500 committed:
chore: make filter not required to submit empty search

wpf500 committed:
chore: upgrade dependencies

wpf500 committed:
fix: type errors

wpf500 committed:
Merge branch 'chore/update-dependencies' into feat/advanced-search

wpf500 committed:
fix: duplicate key and bad import from merge

wpf500 committed:
feat: collapse advanced search when changing segment

wpf500 committed:
feat: add reset button

wpf500 committed:
feat: move paginated result to component and show above contacts table

wpf500 committed:
fix: show no results for empty segments

wpf500 committed:
refactor: upgrade to common 1.9 and merged rule/filter

wpf500 committed:
fix: make a deep copy of rules

wpf500 committed:
feat: only show reset when rules differ

wpf500 committed:
fix: propogate event through proxy component

wpf500 committed:
chore: i18n for nested messages

wpf500 committed:
fix: can reset after completing search

wpf500 committed:
fix: bad input widths and missing i18n

wpf500 committed:
Merge branch 'main' into feat/advanced-search

wpf500 committed:
Merge branch 'main' into feat/advanced-search

aiscarvalho commented:
Great work with the advanced search!

While testing we found that querying by "Active role is not" appears to not be working (we get the same results with "is" and "is not"):

_Active role is Superadmin_ vs _Active role is not Superadmin_ returns the same results
![Screenshot 2022-11-14 at 15-49-21 Community - beabee dev](https://user-images.githubusercontent.com/122730/201706071-6603a3a0-746d-4125-9ea3-a781360dcb3d.png)
![Screenshot 2022-11-14 at 15-49-04 Community - beabee dev](https://user-images.githubusercontent.com/122730/201706058-498f82ff-c8a3-4fca-a1d0-3661773c21d4.png)

_Active role is Member_ vs _Active role is not Member_ returns the same results
![Screenshot 2022-11-14 at 15-51-21 Community - beabee dev](https://user-images.githubusercontent.com/122730/201706083-da7bbc7f-6659-46e6-9b51-f33c6e9b65a0.png)
![Screenshot 2022-11-14 at 15-50-46 Community - beabee dev](https://user-images.githubusercontent.com/122730/201706080-0e07db22-c0d0-427c-a823-362f21a4f01d.png)

And we have two of suggestions to improve the UX:
- Ignore empty querys/rules. Currently the search button is inactive if there's an empty rule and it's a behavior that isn't clear at first:
![image](https://user-images.githubusercontent.com/122730/201706746-71f04d4f-9fcd-40c9-8dc6-ccb5870f4783.png)

- When building a rule for **Member since** the default is _Member since is dd-mm-yyyy_ which translates to _Member since **exact date is** dd-mm-yyyy_.   
Would it be possible to replace _is/is not_ by _exact date is/exact date is not_?  
Alternatively we'd suggest switching the default operator to _after/before_ since it's a bit more intuitive. Same thing for _Contribution cancelled exact date is dd-mm-yyyy_ and _Member until exact date is dd-mm-yyyy_
![image](https://user-images.githubusercontent.com/122730/201708065-18cf6872-d9e6-400c-86bb-890a4a3c9dca.png)

wpf500 commented:
@aiscarvalho I've fixed the "Active role" filter, thanks for pointing it out. All of the filters "Active role", "Active member", "Member since" and "Member expires" will all be replaced soon with a role filter that is much more intuitive.

I've changed "is" to "is on" rather than "exact date is" as I think it conveys the same meaning and slightly cleaner, what do you think?

I can remove the validation on selecting a filter, but it will still cause the button to be disabled when they choose a filter until they fill in the arguments:
![image](https://user-images.githubusercontent.com/2084823/202183944-d7061ebd-4e90-4cc7-83b5-43c1ff7dba3b.png)

I did try this before but Tobi and I found it a bit strange, what should it do when this happens? Still ignore them?
![image](https://user-images.githubusercontent.com/2084823/202183853-879c523e-a871-4d1b-bcdd-c59a5a6dd830.png)

aiscarvalho commented:
> @aiscarvalho I've fixed the "Active role" filter, thanks for pointing it out. All of the filters "Active role", "Active member", "Member since" and "Member expires" will all be replaced soon with a role filter that is much more intuitive.

:+1: 
 
> I've changed "is" to "is on" rather than "exact date is" as I think it conveys the same meaning and slightly cleaner, what do you think?

Perfect!
 
> I can remove the validation on selecting a filter, but it will still cause the button to be disabled when they choose a filter until they fill in the arguments:

Don't remove validation! I think I mis-explained.
In this example, where the filter is almost complete, it's nice to show the input validation. It's a quick way to hint why the "search" button is disabled.

> I did try this before but Tobi and I found it a bit strange, what should it do when this happens? Still ignore them? ![image](https://user-images.githubusercontent.com/2084823/202183853-879c523e-a871-4d1b-bcdd-c59a5a6dd830.png)

Agreed. Then just disregard our suggestion, let's keep it as is :)
